matrix/square_matrix.py

```diff
@@ -64,7 +65,33 @@ def square_root(self):
         raise NotImplementedError
 
     def LU_decomposition(self):
-        raise NotImplementedError
+        """
+        Returns: the LU decomposition of the self.ls_entries matrix
+        self.ls_entries matrix must be square
+            L, U where [A] = [L][U]
+            L: <Matrix> m x m lower triangular matrix
+            U: <Matrix> m x m upper triangular matrix
+        """
+
+        # L and U are both square matrices
+        L = self.identity(self.len_row).ls_entries
+        U = [[0] * self.len_row for x in range(self.len_row)]
+        A = SquareMatrix(copy.deepcopy(self.ls_entries))
+
+        # Need to pivot first for stability
+        for i in range(self.len_row):
+            minor_col_i = list(map(abs, A[i:self.len_row, i]))
+            max_row_index = minor_col_i.index(max(minor_col_i)) + i
+            A[i], A[max_row_index] = A[max_row_index], A[i]
+
+        for i in range(self.len_row):
+            for j in range(i, self.len_row):  # update matrix after pivoting
+                sum_upper = sum([L[i][k] * U[k][j] for k in range(i)])
+                sum_lower = sum([L[j][k] * U[k][i] for k in range(i)])
+                U[i][j] = A[i][j] - sum_upper
+                L[j][i] = (A[j][i] - sum_lower) / U[i][i] if U[i][i] != 0 else 0
+
+        return Matrix(L), Matrix(U)
```

Comment:
Whole pull request looks amazing! I'm ready to merge. Just had one question, should the outputs here be `SquareMatrix(L), SquareMatrix(U)`? 
I'm thinking in the future I'm going to make a Triangular matrix class that is a child of the Square matrix class, so will replace that anyway. 
Might as well merge now and then later make this change.